### PR TITLE
expression: fix incompatible timestamp conversion between mysql and tidb

### LINKED ISSUE #25093: The result of `timestamp` is not compatible with MySQL
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
drop table if exists t;
create table t(col decimal(45,8) default 13.654 not null);
insert  into t set col = 0.4352;

set @@sql_mode="ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
select timestamp(123.123);
select timestamp(0.123);
select timestamp(col) from t;

set @@sql_mode="ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
select timestamp(123.123);
select timestamp(0.123);
select timestamp(col) from t;
```

### 2. What did you expect to see? (Required)
```
mysql>  set @@sql_mode="ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
Query OK, 0 rows affected (0.01 sec)
mysql>  select timestamp(123.123);
+-------------------------+
| timestamp(123.123)      |
+-------------------------+
| 2000-01-23 00:00:00.000 |
+-------------------------+
1 row in set (0.00 sec)
mysql>  select timestamp(0.123);
+------------------+
| timestamp(0.123) |
+------------------+
| NULL             |
+------------------+
1 row in set, 1 warning (0.00 sec)
mysql> select timestamp(col) from t;
+----------------+
| timestamp(col) |
+----------------+
| NULL           |
+----------------+
1 row in set, 1 warning (0.01 sec)


mysql> set @@sql_mode="ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
Query OK, 0 rows affected, 1 warning (0.00 sec)
mysql> select timestamp(123.123);
+-------------------------+
| timestamp(123.123)      |
+-------------------------+
| 2000-01-23 00:00:00.000 |
+-------------------------+
1 row in set (0.01 sec)
mysql> select timestamp(0.123);
+-------------------------+
| timestamp(0.123)        |
+-------------------------+
| 0000-00-00 00:00:00.123 |
+-------------------------+
1 row in set (0.00 sec)
mysql> select timestamp(col) from t;
+----------------------------+
| timestamp(col)             |
+----------------------------+
| 0000-00-00 00:00:00.435200 |
+----------------------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
tidb> set @@sql_mode="ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
Query OK, 0 rows affected (0.00 sec)
tidb> select timestamp(123.123);
+-------------------------+
| timestamp(123.123)      |
+-------------------------+
| 2000-01-23 00:00:00.000 |
+-------------------------+
1 row in set (0.00 sec)
tidb> select timestamp(0.123);
+-------------------------+
| timestamp(0.123)        |
+-------------------------+
| 0000-00-00 00:00:00.000 |
+-------------------------+
1 row in set (0.00 sec)
tidb> select timestamp(col) from t;
+----------------------------+
| timestamp(col)             |
+----------------------------+
| 0000-00-00 00:00:00.000000 |
+----------------------------+
1 row in set (0.00 sec)

tidb> set @@sql_mode="ONLY_FULL_GROUP_BY,STRICT_TRANS_TABLES,ERROR_FOR_DIVISION_BY_ZERO,NO_ENGINE_SUBSTITUTION";
Query OK, 0 rows affected (0.00 sec)
tidb> select timestamp(123.123);
+-------------------------+
| timestamp(123.123)      |
+-------------------------+
| 2000-01-23 00:00:00.000 |
+-------------------------+
1 row in set (0.01 sec)
tidb> select timestamp(0.123);
+-------------------------+
| timestamp(0.123)        |
+-------------------------+
| 0000-00-00 00:00:00.000 |
+-------------------------+
1 row in set (0.00 sec)
tidb> select timestamp(col) from t;
+----------------------------+
| timestamp(col)             |
+----------------------------+
| 0000-00-00 00:00:00.000000 |
+----------------------------+
1 row in set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

0158f71cda6ccbbd382c6e8759904234baca259c

### PR DESCRIPTION
**Pull request by sylzd**
### What problem does this PR solve?

Issue Number: close #25093

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- Fix timestamp convertion case "0.XXX" -> "0000-00-00 00:00:00.XXX".

### CODE REVIEW

**Comment by sylzd** [comment]
@zimulala @XuHuaiyu   PTAL

**Review by XuHuaiyu — COMMENTED** [review]
It seems not right.
We need to take the sql_mode into consideration.
As you can see from the issue, the result differs when there is `NO_ZERO_IN_DATE,NO_ZERO_DATE` in the sql_mode or not.

**Comment by sylzd** [comment]
> It seems not right.
> We need to take the sql_mode into consideration.
> As you can see from the issue, the result differs when there is `NO_ZERO_IN_DATE,NO_ZERO_DATE` in the sql_mode or not.

I don't think the expect result is right,  I've tried the same sql in mysql5.7, as follows:
image

**Comment by sylzd** [comment]
since `NO_ZERO_IN_DATE` and `NO_ZERO_DATE` is deprecated, I think it's resonable to keep the same with mysql5.7.25.

**Comment by sylzd** [comment]
can anyone else help to review? @mmyj @tiancaiamao   I've rebased many times~

**Comment by mmyj** [comment]
@sylzd I will help to review it soon.

**Comment by XuHuaiyu** [comment]
@sylzd I'll finish the review today.

File: types/time.go [hunk 1]
```diff
@@ -1020,6 +1022,14 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int8, isFloat b
 
 			year, month, day, hour, minute, second =
 				dateTime.Year(), dateTime.Month(), dateTime.Day(), dateTime.Hour(), dateTime.Minute(), dateTime.Second()
+			if seps[0] == "0" {
```
Comment (XuHuaiyu) [inline types/time.go:1025]: Add a test case for `sep[0] != "0"`

File: types/time.go [hunk 2]
```diff
@@ -1020,6 +1022,14 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int8, isFloat b
...
+				microsecond, overflow, err = ParseFrac(fracStr, fsp)
+				if err != nil {
+					return ZeroDatetime, errors.Trace(err)
+				}
+				truncatedOrIncorrect = overflow
```
Comment (XuHuaiyu) [inline types/time.go:1030]: Add a test case for `overflow == true`

File: types/time.go [hunk 3]
```diff
@@ -1020,6 +1022,14 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int8, isFloat b
...
+			}
```
Comment (XuHuaiyu) [inline types/time.go:1031]: Add a comment for what does this if block mean

Comment (sylzd) [reply types/time.go:1032]: added

File: types/time.go (hunk 2 above)
Comment (sylzd) [reply types/time.go:1031]: added

File: types/time.go (hunk 1 above)
Comment (sylzd) [reply types/time.go:1026]: added

File: types/time.go [hunk 4]
```diff
@@ -1020,6 +1022,15 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int8, isFloat b
 
 			year, month, day, hour, minute, second =
 				dateTime.Year(), dateTime.Month(), dateTime.Day(), dateTime.Hour(), dateTime.Minute(), dateTime.Second()
+			// Only for case: 0.XXX
+			if seps[0] == "0" {
+				microsecond, overflow, err = ParseFrac(fracStr, fsp)
+				if err != nil {
+					return ZeroDatetime, errors.Trace(err)
+				}
+				truncatedOrIncorrect = overflow
```
Comment (mmyj) [inline types/time.go:1031]: This line looks useless?  This branch is never use `truncatedOrIncorrect`. 

By the way, I think this logic is same as following
```
if seps[0] == "0" || l >= 9 && l <= 14 {
  hhmmss = true
}
```
I had tried it, and passed all the tests.

File: types/time.go (hunk 1 above)
Comment (mmyj) [reply types/time.go:1026]: It seems that you have not added a test for `sep[0] != "0"` branch.

File: types/time.go (hunk 4 above)
Comment (sylzd) [reply types/time.go:1031]: It's not added by me. commit info as follows:
```
commit 8ed9a6ff6d8bb5d9c74f936d01182d4859ab14d1 
Author: Lenshood Date: 2020/10/15, 3:46 下午 
 expression: fix TIMESTAMP func get wrong result with decimal (#15185) (#20088)
```

File: types/time.go (hunk 1 above)
Comment (sylzd) [reply types/time.go:1026]: you're right!  case `1.234` returned by `ParseDatetimeFromNum`.

File: types/time.go (hunk 4 above)
Comment (mmyj) [reply types/time.go:1031]: I mean that the logic of this code
```
switch len(seps) {
case 1:
	if isFloat {
		// do something
		if seps[0] == "0" {
			// do something
			microsecond, overflow, err = ParseFrac(fracStr, fsp)
			if err != nil {
				return ZeroDatetime, errors.Trace(err)
			}
			truncatedOrIncorrect = overflow
		}
		// do something
		if l >= 9 && l <= 14 {
			hhmmss = true
		}
		break
	}
}
// do something
if hhmmss {
	microsecond, overflow, err = ParseFrac(fracStr, fsp)
	if err != nil {
		return ZeroDatetime, errors.Trace(err)
	}
}
```
 is the same as 
```
switch len(seps) {
case 1:
	if isFloat {
		// do something
		if seps[0] == "0" ||l >= 9 && l <= 14 {
			hhmmss = true
		}
		break
	}
}
// do something
if hhmmss {
	microsecond, overflow, err = ParseFrac(fracStr, fsp)
	if err != nil {
		return ZeroDatetime, errors.Trace(err)
	}
}
```

Comment (sylzd) [reply types/time.go]: Yep, maybe the semantic is a little different.

File: types/time.go [hunk 5]
```diff
@@ -1133,8 +1137,6 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int8, isFloat b
 		}
 	}
 
-	var microsecond int
-	var overflow bool
```
Comment (mmyj) [inline types/time.go:1137]: We don't need to move those any more.

Comment (sylzd) [reply types/time.go]: Yep, thanks for your patience.

File: types/time.go [hunk 6]
```diff
@@ -1020,7 +1020,9 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int8, isFloat b
 
 			year, month, day, hour, minute, second =
 				dateTime.Year(), dateTime.Month(), dateTime.Day(), dateTime.Hour(), dateTime.Minute(), dateTime.Second()
-			if l >= 9 && l <= 14 {
+
+			// case: 0.XXX or like "20170118.999"
+			if seps[0] == "0" || l >= 9 && l <= 14 {
```
Comment (mmyj) [inline types/time.go:1025]: Should we use `seps[0] == "0" || (l >= 9 && l <= 14)` here? It will look more clear.

Comment (sylzd) [reply types/time.go]: Exactly!

**Review by mmyj — APPROVED** [review]
lgtm

**Comment by mmyj** [comment]
PTAL @XuHuaiyu

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**